Enhance/#6035 - Reduce code complexity in `SearchFunnelWidget`

hussain-t opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #6035 

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

techanvil commented on assets/js/modules/search-console/components/dashboard/SearchFunnelWidget/index.js:
```diff
@@ -306,14 +226,50 @@ const SearchFunnelWidget = ( { Widget, WidgetReportError } ) => {
-		return select( MODULES_ANALYTICS ).getErrorForSelector( 'getReport', [
-			analyticsVisitorsOverviewAndStatsArgs,
-		] );
+		return (
+			! select( MODULES_ANALYTICS ).hasFinishedResolution( 'getReport', [
+				analyticsOverviewArgs,
+			] ) ||
+			! select( MODULES_ANALYTICS ).hasFinishedResolution( 'getReport', [
+				analyticsStatsArgs,
+			] ) ||
+			! select( MODULES_ANALYTICS ).hasFinishedResolution( 'getReport', [
+				analyticsVisitorsOverviewAndStatsArgs,
+			] ) ||
+			! select( MODULES_ANALYTICS ).hasFinishedResolution(
+				'getGoals',
+				[]
+			)
+		);
```
This works, but we can make it a bit more readable:

```suggestion
		const { hasFinishedResolution } = select( MODULES_ANALYTICS );

		return ! (
			hasFinishedResolution( 'getReport', [ analyticsOverviewArgs ] ) &&
			hasFinishedResolution( 'getReport', [ analyticsStatsArgs ] ) &&
			hasFinishedResolution( 'getReport', [
				analyticsVisitorsOverviewAndStatsArgs,
			] ) &&
			hasFinishedResolution( 'getGoals', [] )
		);
```

techanvil commented on assets/js/modules/search-console/components/dashboard/SearchFunnelWidget/index.js:
```diff
@@ -306,14 +226,50 @@ const SearchFunnelWidget = ( { Widget, WidgetReportError } ) => {
+			return null;
+		}
+
+		return (
+			select( MODULES_ANALYTICS ).getErrorForSelector( 'getReport', [
+				analyticsOverviewArgs,
+			] ) ||
+			select( MODULES_ANALYTICS ).getErrorForSelector( 'getReport', [
+				analyticsStatsArgs,
+			] ) ||
+			select( MODULES_ANALYTICS ).getErrorForSelector( 'getReport', [
+				analyticsVisitorsOverviewAndStatsArgs,
+			] ) ||
+			select( MODULES_ANALYTICS ).getErrorForSelector( 'getGoals', [] )
+		);
```
The same point applies here:

```suggestion
		const { getErrorForSelector } = select( MODULES_ANALYTICS );

		return (
			getErrorForSelector( 'getReport', [ analyticsOverviewArgs ] ) ||
			getErrorForSelector( 'getReport', [ analyticsStatsArgs ] ) ||
			getErrorForSelector( 'getReport', [
				analyticsVisitorsOverviewAndStatsArgs,
			] ) ||
			getErrorForSelector( 'getGoals', [] )
		);
```

hussain-t replied:
Updated ✅

techanvil reviewed (changes requested):
Hi @hussain-t, this certainly works but I think we can improve the readability a bit.

Also, please can you add a bit more detail to the QAB? It should really point out that the loading and error behaviour of the widget should be the main focus here, and give some pointers on how to model the error conditions.

hussain-t commented:
> Hi @hussain-t, this certainly works but I think we can improve the readability a bit.
> 
> Also, please can you add a bit more detail to the QAB? It should really point out that the loading and error behaviour of the widget should be the main focus here, and give some pointers on how to model the error conditions.

Thanks, @techanvil. I have addressed the changes and updated the QAB as well.

techanvil reviewed (approved):
LGTM, nice one @hussain-t!
